src/storage/entry.rs

```diff
@@ -0,0 +1,114 @@
+use crate::storage::Storage;
+
+mod boolean;
+#[cfg(feature = "map")]
+mod map;
+mod option;
+mod singleton;
+
+pub trait OccupiedEntry<'this, K, V> {
+    fn key(&self) -> K;
+    fn get(&self) -> &V;
+    fn get_mut(&mut self) -> &mut V;
+    fn into_mut(self) -> &'this mut V;
+    fn insert(&mut self, value: V) -> V;
+    fn remove(self) -> V;
+}
+
+pub trait VacantEntry<'this, K, V> {
+    fn key(&self) -> K;
+    fn insert(self, value: V) -> &'this mut V;
+}
+
+pub enum Entry<Occupied, Vacant> {
+    Occupied(Occupied),
+    Vacant(Vacant),
+}
+
+impl<Occupied, Vacant> Entry<Occupied, Vacant> {
+    pub fn or_insert<'this, K, V>(self, default: V) -> &'this mut V
+    where
+        Occupied: OccupiedEntry<'this, K, V>,
+        Vacant: VacantEntry<'this, K, V>,
+    {
+        match self {
+            Entry::Occupied(entry) => entry.into_mut(),
+            Entry::Vacant(entry) => entry.insert(default),
+        }
+    }
+
+    pub fn or_insert_with<'this, K, V, F: FnOnce() -> V>(self, default: F) -> &'this mut V
+    where
+        Occupied: OccupiedEntry<'this, K, V>,
+        Vacant: VacantEntry<'this, K, V>,
+    {
+        match self {
+            Entry::Occupied(entry) => entry.into_mut(),
+            Entry::Vacant(entry) => entry.insert(default()),
+        }
+    }
+
+    pub fn or_insert_with_key<'this, K, V, F: FnOnce(K) -> V>(self, default: F) -> &'this mut V
+    where
+        Occupied: OccupiedEntry<'this, K, V>,
+        Vacant: VacantEntry<'this, K, V>,
+    {
+        match self {
+            Entry::Occupied(entry) => entry.into_mut(),
+            Entry::Vacant(entry) => {
+                let value = default(entry.key());
+                entry.insert(value)
+            }
+        }
+    }
+
+    pub fn key<'this, K, V>(&self) -> K
+    where
+        Occupied: OccupiedEntry<'this, K, V>,
+        Vacant: VacantEntry<'this, K, V>,
+    {
+        match self {
+            Entry::Occupied(entry) => entry.key(),
+            Entry::Vacant(entry) => entry.key(),
+        }
+    }
+
+    #[allow(clippy::return_self_not_must_use)]
+    pub fn and_modify<'this, K, V, F: FnOnce(&mut V)>(self, f: F) -> Self
+    where
+        Occupied: OccupiedEntry<'this, K, V>,
+        Vacant: VacantEntry<'this, K, V>,
+    {
+        match self {
+            Entry::Occupied(mut entry) => {
+                f(entry.get_mut());
+                Entry::Occupied(entry)
+            }
+            Entry::Vacant(entry) => Entry::Vacant(entry),
+        }
+    }
+
+    pub fn or_default<'this, K, V>(self) -> &'this mut V
+    where
+        V: Default,
+        Occupied: OccupiedEntry<'this, K, V>,
+        Vacant: VacantEntry<'this, K, V>,
+    {
+        match self {
+            Entry::Occupied(entry) => entry.into_mut(),
+            Entry::Vacant(entry) => entry.insert(Default::default()),
+        }
+    }
+}
+
```

Comment:
I'd like to change this trait to use GATs:

```rust
pub trait StorageEntry<K, V>: Storage<K, V> {
    type Occupied<'this>: OccupiedEntry<'this, K, V>
    where
        Self: 'this;

    type Vacant<'this>: VacantEntry<'this, K, V>
    where
        Self: 'this;

    fn entry(&mut self, key: K) -> Entry<Self::Occupied<'_>, Self::Vacant<'_>>;
}
```

Reply:
What's the benefit? Having it this way drastically simplifies things (see https://github.com/udoprog/fixed-map/commit/68125048a2536ed41b2863bf4026eb746e7d193e) which makes implementing the proc macro waaayy easier. This way is also more similar to the `OccupiedEntry` and `VacantEntry` traits.

Reply:
I'm going offline now. But my high level reasoning is that if we're using GATs it would be using an HRTB, which I'm sure would be associated with `&mut self`. I'm not *entirely* sure if the lifetime we pass in through the trait bound would or could end up being constrained in some other way so that the entry API couldn't be used in some manner. With the current implementation *I don't think so*, but I'm also not sure.

I can look and think closer about it tomorrow if you're reluctant to make the change.

Reply:
So the kind of issue I'm concerned about is expressed in [This Playground](https://play.rust-lang.org/?version=beta&mode=debug&edition=2021&gist=16e5bbd5dea3f8a2b18d35fff609ff5c).

BTW: The fix above is to switch to an HRTB:

```rust
fn test1<T>(mut values: T)
where
    T: for<'a> Values<'a, u32>,
{
    *values.value().into_mut() += 1;
}
```

I'm not sure whether the current formulation of the `fn entry` function on `Map` would trigger *some* kind of lifetime constraint. But I would be more comfortable using GATs since they'd simply allows us to do "the right thing" by default.

I've also tried to move the code over to using GATs as an experiment, and it doesn't make the code harder to deal with IMO: https://github.com/udoprog/fixed-map/commit/4bdf4dcd8cd3b2a5f03b767a2ea7981ca484ec1b

So unless we can reason or prove that we won't have any issues with weirdly constrained lifetimes I'd be much happier moving to GATs.

Reply:
And for the record, I think our use is more akin to this (because we associated the lifetime with an exclusive reference):
```rust
fn test1<'a, T>(values: &'a mut T)
where
    T: Values<'a, u32>,
{
    *values.value().into_mut() += 1;
}
```
Which *doesn't* suffer from the above issue. I just don't know for sure whether it might suffer from some other subtle issue.

Reply:
Your attempt was much more successful than mine, which was painful by comparison. I've gone ahead and pulled in yours, thanks.